File: src/magicpanel/reactions.py
"""Reaction-persistence primitives shared by all scenes
(docs/00-engine.md: "Reaction persistence model").

Four classes of reaction, selected per rule rather than fixed globally:

- transient: reverts to baseline automatically after a fixed duration.
- sticky: persists until an explicit resolving event clears it.
- duration_bound: active only between a start event and its matching end
  event.
- accumulating: adds to durable state (see state.py) rather than
  reverting; only clears on an explicit reset event.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum, auto

from magicpanel.state import AccumulatingStateStore


class ReactionKind(Enum):
    TRANSIENT = auto()
    STICKY = auto()
    DURATION_BOUND = auto()
    ACCUMULATING = auto()


@dataclass(frozen=True)
class ReactionRule:
    name: str
    kind: ReactionKind
    trigger_event: str

    # transient
    duration_seconds: float | None = None
    # sticky
    resolve_event: str | None = None
    # duration_bound
    end_event: str | None = None
    # accumulating
    reset_event: str | None = None


class ReactionEngine:
    """Tracks which reactions are currently active for a scene, driven by
    incoming events and elapsed time.
    """

    def __init__(
        self,
        rules: list[ReactionRule],
        accumulator: AccumulatingStateStore | None = None,
    ) -> None:
        self._rules = rules
        self._accumulator = accumulator or AccumulatingStateStore()

        self._by_trigger: dict[str, ReactionRule] = {}
        self._by_resolve: dict[str, list[ReactionRule]] = {}
        self._by_end: dict[str, list[ReactionRule]] = {}
        self._by_reset: dict[str, list[ReactionRule]] = {}
        for rule in rules:
            self._by_trigger[rule.trigger_event] = rule
            if rule.resolve_event:
                self._by_resolve.setdefault(rule.resolve_event, []).append(rule)
            if rule.end_event:
                self._by_end.setdefault(rule.end_event, []).append(rule)
            if rule.reset_event:
                self._by_reset.setdefault(rule.reset_event, []).append(rule)

        # active[name] = expiry timestamp (transient) or True (sticky/duration_bound)
        self._active: dict[str, float | bool] = {}
        self._clock: float = 0.0

    def handle_event(self, event_name: str) -> None:
        # Clear sticky reactions this event resolves.
        for rule in self._by_resolve.get(event_name, []):
            self._active.pop(rule.name, None)

        # End duration-bound reactions this event closes.
        for rule in self._by_end.get(event_name, []):
            self._active.pop(rule.name, None)

        # Reset accumulating reactions this event clears.
        for rule in self._by_reset.get(event_name, []):
            self._accumulator.reset(rule.name)

        rule = self._by_trigger.get(event_name)
        if rule is None:
            return

        if rule.kind is ReactionKind.TRANSIENT:
            self._active[rule.name] = self._clock + (rule.duration_seconds or 0.0)
        elif rule.kind in (ReactionKind.STICKY, ReactionKind.DURATION_BOUND):
            self._active[rule.name] = True
        elif rule.kind is ReactionKind.ACCUMULATING:
            self._accumulator.increment(rule.name)

    def tick(self, dt: float) -> None:
        self._clock += dt
        expired = [
            name
            for name, expiry in self._active.items()
            if isinstance(expiry, float) and expiry <= self._clock
        ]
        for name in expired:
            del self._active[name]

    def is_active(self, name: str) -> bool:
        return name in self._active

    def accumulated(self, name: str) -> int:
        return self._accumulator.get(name)

File: cpp/src/reactions.cpp
#include "magicpanel/reactions.h"

#include <algorithm>

namespace magicpanel {

ReactionEngine::ReactionEngine(std::vector<ReactionRule> rules,
                               AccumulatingStateStore* accumulator)
    : rules_(std::move(rules)), accumulator_(accumulator) {
  for (std::size_t i = 0; i < rules_.size(); ++i) {
    index_rule(i);
  }
}

void ReactionEngine::index_rule(std::size_t i) {
  auto const& rule = rules_[i];
  by_trigger_[rule.trigger_event].push_back(i);
  if (!rule.resolve_event.empty()) {
    by_resolve_[rule.resolve_event].push_back(i);
  }
  if (!rule.end_event.empty()) {
    by_end_[rule.end_event].push_back(i);
  }
  if (!rule.reset_event.empty()) {
    by_reset_[rule.reset_event].push_back(i);
  }
}

void ReactionEngine::clear_active(std::string const& name) {
  active_.erase(name);
  transient_expiry_.erase(name);
}

void ReactionEngine::handle_event(std::string const& event_name) {
  if (auto it = by_resolve_.find(event_name); it != by_resolve_.end()) {
    for (std::size_t index : it->second) {
      clear_active(rules_[index].name);
    }
  }

  if (auto it = by_end_.find(event_name); it != by_end_.end()) {
    for (std::size_t index : it->second) {
      clear_active(rules_[index].name);
    }
  }

  if (accumulator_ != nullptr) {
    if (auto it = by_reset_.find(event_name); it != by_reset_.end()) {
      for (std::size_t index : it->second) {
        accumulator_->reset(rules_[index].name);
      }
    }
  }

  auto triggered = by_trigger_.find(event_name);
  if (triggered == by_trigger_.end()) {
    return;
  }

  for (std::size_t index : triggered->second) {
    auto const& rule = rules_[index];
    switch (rule.kind) {
      case ReactionKind::Transient:
        active_.insert(rule.name);
        transient_expiry_[rule.name] = clock_ + rule.duration_seconds;
        break;
      case ReactionKind::Sticky:
      case ReactionKind::DurationBound:
        active_.insert(rule.name);
        break;
      case ReactionKind::Accumulating:
        if (accumulator_ != nullptr) {
          accumulator_->increment(rule.name);
        }
        break;
    }
  }
}

void ReactionEngine::tick(float dt) {
  clock_ += dt;
  std::vector<std::string> expired;
  for (auto const& [name, expiry] : transient_expiry_) {
    if (expiry <= clock_) {
      expired.push_back(name);
    }
  }
  for (auto const& name : expired) {
    clear_active(name);
  }
}

bool ReactionEngine::is_active(std::string const& name) const {
  return active_.find(name) != active_.end();
}

int ReactionEngine::accumulated(std::string const& name) const {
  return accumulator_ == nullptr ? 0 : accumulator_->get(name);
}

}  // namespace magicpanel
